core/src/test/java/com/linecorp/armeria/client/HttpClientResponseTimeoutHandlerTest.java

```diff
@@ -0,0 +1,162 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.awaitility.Awaitility.await;
+
+import java.time.Duration;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.EnumSource;
+
+import com.linecorp.armeria.common.ClosedSessionException;
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpRequestWriter;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpResponseWriter;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.logging.RequestLog;
+import com.linecorp.armeria.common.logging.RequestLogAvailability;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.testing.junit.server.ServerExtension;
+
+class HttpClientResponseTimeoutHandlerTest {
+
+    @RegisterExtension
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(final ServerBuilder sb) throws Exception {
+            sb.service("/slow", (ctx, req) -> {
+                final HttpResponseWriter response = HttpResponse.streaming();
+                response.write(ResponseHeaders.of(HttpStatus.OK));
+                response.write(HttpData.ofUtf8("slow response"));
+                return response;
+            });
+        }
+    };
+
+    @ParameterizedTest
+    @CsvSource({
+            "H1C, true", "H1C, false",
+            "H2C, true", "H2C, false"
+    })
+    void testResponseTimeoutHandler(SessionProtocol protocol, boolean useResponseTimeoutHandler) {
+        final AtomicReference<RequestLog> logHolder = new AtomicReference<>();
+        final IllegalStateException reqCause = new IllegalStateException("abort request");
+        final AtomicBoolean invokeResponseTimeoutHandler = new AtomicBoolean(false);
+        final HttpClient client = HttpClient.builder(server.uri(protocol, "/"))
+                .responseTimeout(Duration.ofSeconds(2))
+                .decorator((delegate, ctx, req) -> {
+                    if (useResponseTimeoutHandler) {
+                        ctx.setResponseTimeoutHandler(() -> {
+                            ctx.request().abort(reqCause);
```

Comment:
Could we provide two tests? One that aborts a request and the other that aborts a response.

Reply:
Ah, missed that one. Never mind.